core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/AvroUpsertMonitor.scala

```diff
@@ -0,0 +1,234 @@
+                                  googleServicesDAO: GoogleServicesDAO,
+                                  samDAO: SamDAO,
+                                  googleStorage: GoogleStorageService[IO],
+                                  pubSubDao: GooglePubSubDAO,
+                                  pubSubSubscriptionName: String,
+                                  avroUpsertBucketName: String,
+                                  batchSize: Int)(implicit cs: ContextShift[IO])
+  extends Actor
+    with LazyLogging
+    with FutureSupport {
+  import AvroUpsertMonitor._
+  import context._
+
+  case class AvroMetadataJson(namespace: String, name: String, userSubjectId: String, userEmail: String, jobId: String)
+  implicit val avroMetadataJsonFormat = jsonFormat5(AvroMetadataJson)
+
+  self ! StartMonitorPass
+
+  // fail safe in case this actor is idle too long but not too fast (1 second lower limit)
+  setReceiveTimeout(max((pollInterval + pollIntervalJitter) * 10, 1 second))
+
+  private def max(durations: FiniteDuration*): FiniteDuration = {
+    implicit val finiteDurationIsOrdered = scala.concurrent.duration.FiniteDuration.FiniteDurationIsOrdered
+    durations.max
+  }
+
+  override def receive = {
+    case StartMonitorPass =>
+      // start the process by pulling a message and sending it back to self
+      pubSubDao.pullMessages(pubSubSubscriptionName, 1).map(_.headOption) pipeTo self
+
+    case Some(message: PubSubMessage) =>
+      //we received a message, so we will parse it and try to upsert it
+      logger.info(s"received avro upsert message: $message")
+      val (jobId, file) = parseMessage(message)
+
+      file match {
+        case "upsert.json" => initAvroUpsert(jobId, message.ackId).map(_ => ImportComplete) pipeTo self
+        case _ => acknowledgeMessage(message.ackId).map(_ => None) pipeTo self //some other file (i.e. success/failure log, metadata.json)
+      }
+
+    case None =>
+      // there was no message so wait and try again
+      val nextTime = org.broadinstitute.dsde.workbench.util.addJitter(pollInterval, pollIntervalJitter)
+      system.scheduler.scheduleOnce(nextTime, self, StartMonitorPass)
+
+    case ImportComplete => self ! None
+
+    case Status.Failure(t) => throw t
+
+    case ReceiveTimeout => throw new WorkbenchException("AvroUpsertMonitorActor has received no messages for too long")
+
+    case x =>
+      logger.info(s"unhandled $x")
+      self ! None
+  }
+
+  private def initAvroUpsert(jobId: String, ackId: String): Future[Unit] = {
+    //start this Future asap because it can be long-running
+    val readUpsertObjectFuture = readUpsertObject(s"$jobId/upsert.json")
+
+    for {
+      avroMetadataJson <- readMetadataObject(s"$jobId/metadata.json")
+      //ack the response after we load the json into memory. pro: don't have to worry about ack timeouts for long operations, con: if someone restarts rawls here the uspert is lost
```

Comment:
Another con: we won't retry if there's a temporary blip with setting up the pet or loading the upsert JSON.

Reply:
agree - I suggest doing the Sam + Google calls first to get the userinfo; if those succeed, start the read of upsert.json, ack the message, perform the upserts. This does incur a delay before starting to read upsert.json, but it protects against and Sam-related problems.